test/unit_tests/braket/aws/test_aws_device.py

```diff
@@ -848,3 +851,105 @@ def test_get_devices_simulators_only(mock_copy_session, aws_session):
+
+    monday_device = MockDevice("ONLINE", ("Monday", "07:00", "17:00"))
+    tuesday_device = MockDevice("ONLINE", ("Tuesday", "07:00", "17:00"))
+    wednesday_device = MockDevice("ONLINE", ("Wednesday", "07:00", "17:00"))
+    thursday_device = MockDevice("ONLINE", ("Thursday", "07:00", "17:00"))
+    friday_device = MockDevice("ONLINE", ("Friday", "07:00", "17:00"))
+    saturday_device = MockDevice("ONLINE", ("Saturday", "07:00", "17:00"))
+    sunday_device = MockDevice("ONLINE", ("Sunday", "07:00", "17:00"))
+    monday_friday_device = MockDevice(
+        "ONLINE", ("Monday", "07:00", "17:00"), ("Friday", "07:00", "17:00")
+    )
+
+    test_devices = (
+        midday_everyday_device,
+        midday_weekday_device,
+        midday_weekend_device,
+        evening_everyday_device,
+        evening_weekday_device,
+        evening_weekend_device,
+        monday_device,
+        tuesday_device,
+        wednesday_device,
+        thursday_device,
+        friday_device,
+        saturday_device,
+        sunday_device,
+        monday_friday_device,
+    )
+
+    test_set = (
+        (datetime(2021, 12, 6, 5, 0, 0), (0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0)),
+        (datetime(2021, 12, 6, 10, 0, 0), (1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1)),
+        (datetime(2021, 12, 6, 20, 0, 0), (0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0)),
+        (datetime(2021, 12, 7, 5, 0, 0), (0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0)),
+        (datetime(2021, 12, 7, 10, 0, 0), (1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0)),
+        (datetime(2021, 12, 7, 20, 0, 0), (0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0)),
+        (datetime(2021, 12, 8, 5, 0, 0), (0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0)),
+        (datetime(2021, 12, 8, 10, 0, 0), (1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0)),
+        (datetime(2021, 12, 8, 20, 0, 0), (0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0)),
+        (datetime(2021, 12, 9, 5, 0, 0), (0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0)),
+        (datetime(2021, 12, 9, 10, 0, 0), (1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0)),
+        (datetime(2021, 12, 9, 20, 0, 0), (0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0)),
+        (datetime(2021, 12, 10, 5, 0, 0), (0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0)),
+        (datetime(2021, 12, 10, 10, 0, 0), (1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1)),
+        (datetime(2021, 12, 10, 20, 0, 0), (0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0)),
+        (datetime(2021, 12, 11, 5, 0, 0), (0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0)),
+        (datetime(2021, 12, 11, 10, 0, 0), (1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0)),
+        (datetime(2021, 12, 11, 20, 0, 0), (0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0)),
+        (datetime(2021, 12, 12, 5, 0, 0), (0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0)),
+        (datetime(2021, 12, 12, 10, 0, 0), (1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0)),
+        (datetime(2021, 12, 12, 20, 0, 0), (0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0)),
+    )
+
+    for test_item in test_set:
```

Comment:
Could we use something like pytest.mark.parametrize ? The problem with the current approach is that if a certain case fails it's hard to tell which one it was.

Reply:
I attempted to move the result table to a parameterized set. It achieved the desired goal of making failures easier to understand, but actually made the code less readable. This is because the expected value table preceded the definition of the devices.

To achieve the desired goal of making test failures understandable and (hopefully) improving readability, I reworked the test case to format an explanation string on assert failures that provides the device, date, expected, and actual values.

https://github.com/aws/amazon-braket-sdk-python/blob/f27e1ce9d563137362714a0602b5e1ca7c7280d5/test/unit_tests/braket/aws/test_aws_device.py#L985-L987

In addition, I broke the test cases into three groups - the always on/always off devices, the weekend/weekday devices, and the named devices. This makes the binary expected value table easier to read as there are fewer columns per set.

https://github.com/aws/amazon-braket-sdk-python/blob/f27e1ce9d563137362714a0602b5e1ca7c7280d5/test/unit_tests/braket/aws/test_aws_device.py#L881-L898